hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -342,36 +263,40 @@ func assignArrayFromArray(
 		return nil
 	}
 
-	return func(reader dst.Expr, writer dst.Expr, generationContext *CodeGenerationContext) []dst.Stmt {
+	return func(reader dst.Expr, writer func(dst.Expr) []dst.Stmt, generationContext *CodeGenerationContext) []dst.Stmt {
 		// We create three obviously related identifiers to use for the conversion
-		id := sourceEndpoint.CreateLocal()
-		itemId := id + "Item"
-		indexId := id + "Index"
-		tempId := id + "List"
+		itemId := sourceEndpoint.CreateLocal("Item")
+		indexId := sourceEndpoint.CreateLocal("Index")
+		tempId := sourceEndpoint.CreateLocal("List")
 
 		declaration := astbuilder.SimpleAssignment(
 			dst.NewIdent(tempId),
 			token.DEFINE,
 			astbuilder.MakeList(dstArray.AsType(generationContext), astbuilder.CallFunc("len", reader)))
 
-		body := conversion(
-			dst.NewIdent(itemId),
-			&dst.IndexExpr{
-				X:     dst.NewIdent(tempId),
-				Index: dst.NewIdent(indexId),
-			},
-			generationContext,
-		)
+		writeToElement := func(expr dst.Expr) []dst.Stmt {
+			return []dst.Stmt{
+				astbuilder.SimpleAssignment(
+					&dst.IndexExpr{
+						X:     dst.NewIdent(tempId),
+						Index: dst.NewIdent(indexId),
+					},
+					token.ASSIGN,
+					expr),
+			}
+		}
 
-		assign := astbuilder.SimpleAssignment(writer, token.ASSIGN, dst.NewIdent(tempId))
+		avoidAliasing := astbuilder.SimpleAssignment(dst.NewIdent(itemId), token.DEFINE, dst.NewIdent(itemId))
```

Comment:
You mentioned in the PR description: 
> Modify both Array and Map generation to create a local shadow of the loop variable to avoid aliasing errors during conversions

Can you expand on how exactly that becomes an issue? My understanding was that as long as you're not doing any capturing lambdas there's no need to shadow. I looked through the golden files tests and I didn't find any instance where you actually needed to do a shadow (although I'm not a compiler so maybe I missed it).

It does seem to result in a decent amount of "awkward" code where the assignment is clearly not required, such as:
```
// Shadow the loop variable to avoid aliasing
nastyValue1 := nastyValue1
nasty := nastyValue1
```

Reply:
If you don't shadow the loop variable and take it's address, dereferencing the pointer will always see the current value.

In PR #429 I modify the AddressOff use to automatically introduce the local if one is needed, so the awkwardness of the code goes away.

I could have also removed the automatic shadowing of the array & map code generation, but I like how it makes the code generation more robust. If a future maintainer inadventently directly takes the address of the loop variable, it will still be safe.